pkg/services/ngalert/models/testing.go

```diff
@@ -20,320 +19,477 @@ import (
+			OkErrState,
 		}
+		return s[g.Rand.Intn(len(s))]
+	}
 
-		for _, mutator := range mutators {
-			mutator(rule)
+	interval := (g.Rand.Int63n(6) + 1) * 10
+	forInterval := time.Duration(interval*g.Rand.Int63n(6)) * time.Second
+
+	var annotations map[string]string = nil
+	if g.Rand.Int63()%2 == 0 {
+		annotations = GenerateAlertLabels(g.Rand.Intn(5), "ann-")
+	}
+	var labels map[string]string = nil
+	if g.Rand.Int63()%2 == 0 {
+		labels = GenerateAlertLabels(g.Rand.Intn(5), "lbl-")
+	}
+
+	var dashUID *string = nil
+	var panelID *int64 = nil
+	if g.Rand.Int63()%2 == 0 {
+		d := util.GenerateShortUID()
+		dashUID = &d
+		p := g.Rand.Int63n(1500)
+		panelID = &p
+	}
+
+	var ns []NotificationSettings
+	if g.Rand.Int63()%2 == 0 {
+		ns = append(ns, NotificationSettingsGen()())
+	}
+
+	rule := AlertRule{
+		ID:                   0,
+		OrgID:                g.Rand.Int63n(1500) + 1, // Prevent OrgID=0 as this does not pass alert rule validation.
+		Title:                fmt.Sprintf("title-%s", util.GenerateShortUID()),
+		Condition:            "A",
+		Data:                 []AlertQuery{g.GenerateQuery()},
+		Updated:              time.Now().Add(-time.Duration(g.Rand.Intn(100) + 1)),
+		IntervalSeconds:      g.Rand.Int63n(60) + 1,
+		Version:              g.Rand.Int63n(1500), // Don't generate a rule ID too big for postgres
+		UID:                  util.GenerateShortUID(),
+		NamespaceUID:         util.GenerateShortUID(),
+		DashboardUID:         dashUID,
+		PanelID:              panelID,
+		RuleGroup:            fmt.Sprintf("group-%s,", util.GenerateShortUID()),
+		RuleGroupIndex:       g.Rand.Intn(1500),
+		NoDataState:          randNoDataState(),
+		ExecErrState:         randErrState(),
+		For:                  forInterval,
+		Annotations:          annotations,
+		Labels:               labels,
+		NotificationSettings: ns,
+	}
+
+	for _, mutator := range g.mutators {
+		mutator(g, &rule)
+	}
+	return rule
+}
+
+func (g *AlertRuleGenerator) GenerateRef() *AlertRule {
+	r := g.Generate()
+	return &r
+}
+
+func (g *AlertRuleGenerator) getCount(bounds ...int) int {
+	count := 0
+	if len(bounds) == 0 {
+		count = g.Rand.Intn(5) + 1
+	}
+	if len(bounds) == 1 {
+		count = bounds[0]
+	}
+	if len(bounds) == 2 {
+		if bounds[0] > bounds[1] {
+			panic("min should not be greater than max")
+		} else if bounds[0] < bounds[1] {
+			count = g.Rand.Intn(bounds[1]-bounds[0]) + bounds[0]
+		} else {
+			count = bounds[0]
 		}
-		return rule
 	}
+	if len(bounds) > 2 {
+		panic("invalid number of parameter must be up to 2")
+	}
+	return count
+}
+
+func (g *AlertRuleGenerator) GenerateMany(bounds ...int) []AlertRule {
+	count := g.getCount(bounds...)
+	result := make([]AlertRule, 0, count)
+	for i := 0; i < count; i++ {
+		result = append(result, g.Generate())
+	}
+	return result
 }
 
-func WithNotEmptyLabels(count int, prefix string) AlertRuleMutator {
-	return func(rule *AlertRule) {
+func (g *AlertRuleGenerator) GenerateManyRef(bounds ...int) []*AlertRule {
```

Comment:
I'm not sure we should keep this bounding behavior, what does this accomplish that the original `rand.Intn()` calls didn't?

Reply:
It just simplifies the syntax.